### DISCUSSION TITLE
Why attributes over interfaces?

### DISCUSSION THREAD

**Discussion by ghost** [symfony/symfony#48539] [Q&A] [answered]
Hi. I am very interested why attributes are now preferred over interfaces and the latter are replaced by the former. The examples are e.g. `MessageHandlerInterface` and `MessageSubscriberInterface` replaced by `AsMessageHandler`, `EventSubscriberInterface` that may be replaced by (multiple) `AsEventListener`. 
I just wonder what are the motivations behind such decision. Why attributes are better? If I use custom marker interfaces in my codebase, should I replace them with custom Attributes?

Thank you for your explanation.

**Comment by MatTheCat** [comment, marked as answer]
I'm not behind this decision but using interfaces as metadata is kind of a hack (because it does not define any contract), so when you have a tool like attributes to add metadata it's better to use it!

**Reply by kbond** [reply]
Agreed, an empty interface as some kind of _marker_ is a sign to use an attribute instead (imo).

One thing I've wondered about is: is there a significant perf boost to using instanceof  (for interfaces) vs reflection (for attributes).
